Add blog post for job tracking with finalizers GA (#37438)
kubernetes/website: content/en/blog/_posts/2022-12-29-scalable-job-tracking-ga/index.md

---
layout: blog

> title: "Scalable Job tracking goes GA to support massively parallel batch workloads"

Reviewer: ```suggestion
title: "Kubernetes 1.26: Job Tracking, To Support Massively Parallel Batch Workloads, is Generally Available"
```
or
```suggestion
title: "Kubernetes 1.26: Job Tracking is Generally Available"
```

Author: Used the first option, with APA style

date: 2022-12-29
slug: "scalable-job-tracking-ga"
---

**Authors:** Aldo Culquicondor (Google)

> The Kubernetes 1.26 release includes a stable implementation of the Job

Reviewer: ```suggestion
The Kubernetes 1.26 release includes a stable implementation of the 
[Job](/docs/concepts/workloads/controllers/job/)
```
?

Author: Added.

controller that can reliably track a large amount of Jobs with high levels of
parallelism. [SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps)
and [WG Batch](https://github.com/kubernetes/community/tree/master/wg-batch)
have worked on this foundational improvement since Kubernetes 1.22. After
multiple iterations and scale verifications, this is now the default
implementation of the Job controller.

Paired with the Indexed [completion mode](/docs/concepts/workloads/controllers/job/#completion-mode),
the Job controller can handle massively parallel batch Jobs, supporting up to
100k concurrent Pods.

The new implementation also made possible the development of [Pod failure policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy),
which is in beta in the 1.26 release.

## How do I use this feature?

> To use Job tracking with finalizers, upgrade to Kubernetes 1.25 or newer and
> create new Jobs. You can also use this feature in 1.23 and 1.24, if you have the
> ability to enable the `JobTrackingWithFinalizers` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)).

Reviewer: ```suggestion
To use Job tracking with finalizers, upgrade to Kubernetes 1.25 or newer and
create new Jobs. You can also use this feature in 1.23 and 1.24, if you have the
ability to enable the `JobTrackingWithFinalizers` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)).

If your cluster runs Kubernetes 1.26, finalizers for Jobs is a stable feature. For
v1.25, it's behind that feature gate, and your cluster administrators may have
explicitly disabled it - for example, if you have a policy of not using
beta features.
```
?

Author: sg

Jobs created before the upgrade will still be tracked using the legacy behavior.
This is to avoid retroactively adding finalizers to running Pods, which might
introduce race conditions.

> For the maximum performance on large Jobs, we recommend using the

Reviewer: ```suggestion
For the maximum performance on large Jobs, the Kubernetes project recommends using the
```

Author: done

[Indexed completion mode](/docs/concepts/workloads/controllers/job/#completion-mode).
In this mode, the control plane is able to track Job progress with less API
calls.

If you are a developer of operator(s) for batch, [HPC](https://en.wikipedia.org/wiki/High-performance_computing),
(AI)[https://en.wikipedia.org/wiki/Artificial_intelligence], [ML](https://en.wikipedia.org/wiki/Machine_learning)
or related workloads, we encourage you to use the Job API to delegate accurate
progress tracking to Kubernetes. If there is something missing in the Job API

> that forces you to manage plain Pods, we would welcome your feedback and
> contributions in the [Working Group Batch](https://github.com/kubernetes/community/tree/master/wg-batch).

Reviewer: ```suggestion
that forces you to manage plain Pods, the
[Batch Working Group](https://github.com/kubernetes/community/tree/master/wg-batch)
(WG Batch) would welcome your feedback and contributions.
```

Author: Done

this by adding an in-memory cache for removed finalizers.

Still, we faced some issues during the beta stage, leaving some pods stuck
with finalizers in some conditions ([#108645](https://github.com/kubernetes/kubernetes/issues/108645),
[#109485](https://github.com/kubernetes/kubernetes/issues/109485)), and
[#111646](https://github.com/kubernetes/kubernetes/pull/111646)). As a result,
we decided to disable the feature in the 1.23 and 1.24 releases.

Once resolved, we re-enabled the feature for the 1.25 release. Since then, we
have received reports from our customers running tens of thousands of Pods at a
time in their clusters through the Job API. With this success, we decided to
graduate the feature to stable in 1.26, as part of our long term commitment to
make the Job API the best way to run large batch Jobs in a Kubernetes cluster.

To learn more about the feature, you can read the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods).

## Acknowledgments

As in any Kubernetes feature, multiple people contributed to get this done,
from testing and filing bugs to reviewing code.

> I would like to specially thank Jordan Liggitt (Google) for helping me debug

Reviewer: The blog article has just **one** author, so how about:
“On behalf of SIG Apps, I would...”

(or whatever SIG is appropriate).

> I would like to specially thank Jordan Liggitt (Google) for helping me debug

Reviewer 2: That actually works for me.

> and brainstorm solutions for more than one race condition. And Maciej Szulik

Author: Done

---
layout: blog
title: "Kubernetes 1.26: Job Tracking, to Support Massively Parallel Batch Workloads, Is Generally Available"
date: 2022-12-29
slug: "scalable-job-tracking-ga"
---

**Authors:** Aldo Culquicondor (Google)

The Kubernetes 1.26 release includes a stable implementation of the [Job](/docs/concepts/workloads/controllers/job/)
controller that can reliably track a large amount of Jobs with high levels of
parallelism. [SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps)
and [WG Batch](https://github.com/kubernetes/community/tree/master/wg-batch)
have worked on this foundational improvement since Kubernetes 1.22. After
multiple iterations and scale verifications, this is now the default
implementation of the Job controller.

Paired with the Indexed [completion mode](/docs/concepts/workloads/controllers/job/#completion-mode),
the Job controller can handle massively parallel batch Jobs, supporting up to
100k concurrent Pods.

The new implementation also made possible the development of [Pod failure policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy),
which is in beta in the 1.26 release.

## How do I use this feature?

> To use Job tracking with finalizers, upgrade to Kubernetes 1.25 or newer and

Reviewer 2: Would it make sense to link this here?
```suggestion
To use [Job tracking with finalizers](/docs/concepts/workloads/controllers/job/#job-tracking-with-finalizers), upgrade to Kubernetes 1.25 or newer and
```

Author: There is nothing in that doc that is not already explained here and we intend to remove the section in 1.28. So I think it's not worth linking.

Reviewer: Fair enough.

Reviewer 2: Alright!

create new Jobs. You can also use this feature in v1.23 and v1.24, if you have the
ability to enable the `JobTrackingWithFinalizers` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)).

If your cluster runs Kubernetes 1.26, Job tracking with finalizers is a stable
feature. For v1.25, it's behind that feature gate, and your cluster administrators may have
explicitly disabled it - for example, if you have a policy of not using
beta features.

Jobs created before the upgrade will still be tracked using the legacy behavior.
This is to avoid retroactively adding finalizers to running Pods, which might
introduce race conditions.

> For the maximum performance on large Jobs, the Kubernetes project recommends

Reviewer 2: nit.
```suggestion
For maximum performance on large Jobs, the Kubernetes project recommends
```

Author: Done

using the [Indexed completion mode](/docs/concepts/workloads/controllers/job/#completion-mode).
In this mode, the control plane is able to track Job progress with less API
calls.

If you are a developer of operator(s) for batch, [HPC](https://en.wikipedia.org/wiki/High-performance_computing),
(AI)[https://en.wikipedia.org/wiki/Artificial_intelligence], [ML](https://en.wikipedia.org/wiki/Machine_learning)
or related workloads, we encourage you to use the Job API to delegate accurate
progress tracking to Kubernetes. If there is something missing in the Job API
that forces you to manage plain Pods, the [Working Group Batch](https://github.com/kubernetes/community/tree/master/wg-batch)

> would welcome your feedback and contributions.

Reviewer 2: nit.
```suggestion
welcomes your feedback and contributions.
```

Reviewer 3: هلا

Author: Done

### Deprecation notices

During the development of the feature, the control plane added the annotation

> `batch.kubernetes.io/job-tracking` to Jobs created when the feature is enabled.

Author: Done

This allowed a safe transition for older Jobs, but it was never meant to stay.

In the 1.26 release, we deprecated the annotation `batch.kubernetes.io/job-tracking`
and the control plane will stop adding it in Kubernetes 1.27.
Along with that change, we will remove the legacy Job tracking implementation.
As a result, the job controller will track all Jobs using finalizers and it will

> ignore Pods that don't have a the aforementioned finalizer.

Reviewer 2: nit.
```suggestion
ignore Pods that don't have the aforementioned finalizer.
```

Before you upgrade your cluster to 1.27, we recommend that you verify that there

> are no running Jobs that don't have the annotation. Otherwise, you might observe

Reviewer 2: Since the annotation is deprecated, I assume "don't" is unnecessary in this sentence.
```suggestion
are no running Jobs that have the annotation. Otherwise, you might observe
```

Reviewer: The original wording is right, I think. The idea is to **ensure** that no Jobs have the deprecated annotation.

Reviewer 2: To me, the original reads similar to this (in my head)-  There shouldn't be a running job observed that doesn't have the annotation. 

With my suggestion (in my head, again) - There shouldn't be a running job observed that has the annotation.

I think it is probably because I'm not a native English speaker.

Author: I added a couple of sentences for extra context.

in order to keep track of the status.

This dependency made the tracking of Job status unreliable, because Pods can be
deleted from the API for a number of reasons, including:
- The garbage collector removing orphan Pods when a Node goes down.
- The garbage collector removing terminated Pods when they reach a threshold.
- The Kubernetes scheduler preempting a Pod to accomodate higher priority Pods.
- The taint manager evicting a Pod that doesn't tolerate a `NoExecute` taint.
- External controllers, not included as part of Kubernetes, or humans deleting
  Pods.

### The new implementation

When a controller needs to take an action on objects before they are removed, it
should add a [finalizer](/docs/concepts/overview/working-with-objects/finalizers/)
to the objects that it manages.
A finalizer prevents the objects from being deleted from the API until the
finalizers are removed. Once the controller is done doing cleanup and accounting
for the deleted object, it can remove the finalizer from the object and the
control plane removes the object from the API.

This is what the new Job controller is doing: adding a finalizer during Pod

> creation, and removing it after the Pod terminates and is accounted for in the

Reviewer 2: nit.
```suggestion
creation, and removing it after the Pod terminates and accounting it in the
```

Reviewer: To me, the original wording is much more idiomatic.

Reviewer 2: I agree that it sounds more idiomatic. But the verb tenses don't agree, which is why I labelled it as a nit. We're talking about what the new Job controller is doing & if we could phrase the accounting verb to match the other tenses, it'd be great.

Reviewer: The clause “the Pod terminates and is accounted for” makes sense. We could write “the Pod has terminated and has been accounted for” for clarity. How about that?

Job status. However, it wasn't that simple.

The main challenge is that there are at least two objects involved: the Pod
and the Job. While the finalizer lives in the Pod object, the accounting lives

> in the Job object. There is no mechanism to atomically remove the finalizer in

Reviewer 2: Do we mean unchangeably remove the finalizer or automatically remove the finalizer here since atomically = unchangeable, indivisible, or infinitesimal?
```suggestion
in the Job object. There is no mechanism to atomically remove the finalizer in
```

Reviewer 2: Ah okay, that clarifies! Thanks :) I confused it with the literal definition.

the Pod and update the counters in the Job status. And there could be more than
one terminated Pod at a given time.

To solve this problem, we implemented a three staged approach, each translating
to an API call.
1. Add the terminated Pod(s) UID into temporary lists in the Job status
   ([.status.uncountedTerminatedPods](/docs/reference/kubernetes-api/workload-resources/job-v1/#JobStatus)).
2. Remove the finalizer from the Pods(s).
3. Atomically remove UIDs from the temporary lists and increment the `succeeded`
   and `failed` counters in the Job status.

Additional complications come from the fact that the Job controller might
receive the results of the API changes in steps 1 and 2 out of order. We solved
this by adding an in-memory cache for removed finalizers.

Still, we faced some issues during the beta stage, leaving some pods stuck
with finalizers in some conditions ([#108645](https://github.com/kubernetes/kubernetes/issues/108645),

> [#109485](https://github.com/kubernetes/kubernetes/issues/109485)), and

Reviewer 2: ```suggestion
[#109485](https://github.com/kubernetes/kubernetes/issues/109485), and
```

Reviewer: The closing paren matches an earlier opening paren.

Reviewer 2: Are we sure about that @sftim? The subsequent line has an extra closer paren which is why I removed this one.

Reviewer: Ah, yep.

Reviewer 2: Ah okay, I see that there is an and in between. The next line does have an extra close paren. Thanks for the catch!

> [#111646](https://github.com/kubernetes/kubernetes/pull/111646)). As a result,

Reviewer 2: nit. Extra paren. The open paren has been closed in the previous line. 
```suggestion
[#111646](https://github.com/kubernetes/kubernetes/pull/111646). As a result,
```

Reviewer: @alculquicondor I'm not sure which analysis is right, but please do check that what renders here match what you intended.

slug: "scalable-job-tracking-ga"
---

**Authors:** Aldo Culquicondor (Google)

The Kubernetes 1.26 release includes a stable implementation of the [Job](/docs/concepts/workloads/controllers/job/)
controller that can reliably track a large amount of Jobs with high levels of
parallelism. [SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps)
and [WG Batch](https://github.com/kubernetes/community/tree/master/wg-batch)
have worked on this foundational improvement since Kubernetes 1.22. After
multiple iterations and scale verifications, this is now the default
implementation of the Job controller.

Paired with the Indexed [completion mode](/docs/concepts/workloads/controllers/job/#completion-mode),
the Job controller can handle massively parallel batch Jobs, supporting up to
100k concurrent Pods.

The new implementation also made possible the development of [Pod failure policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy),
which is in beta in the 1.26 release.

## How do I use this feature?

To use Job tracking with finalizers, upgrade to Kubernetes 1.25 or newer and
create new Jobs. You can also use this feature in v1.23 and v1.24, if you have the

> ability to enable the `JobTrackingWithFinalizers` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)).

Reviewer: ```suggestion
ability to enable the `JobTrackingWithFinalizers` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
```

Author: Oops

If your cluster runs Kubernetes 1.26, Job tracking with finalizers is a stable
feature. For v1.25, it's behind that feature gate, and your cluster administrators may have
explicitly disabled it - for example, if you have a policy of not using
beta features.

Jobs created before the upgrade will still be tracked using the legacy behavior.
This is to avoid retroactively adding finalizers to running Pods, which might
introduce race conditions.

For maximum performance on large Jobs, the Kubernetes project recommends
using the [Indexed completion mode](/docs/concepts/workloads/controllers/job/#completion-mode).
In this mode, the control plane is able to track Job progress with less API
calls.

If you are a developer of operator(s) for batch, [HPC](https://en.wikipedia.org/wiki/High-performance_computing),

> (AI)[https://en.wikipedia.org/wiki/Artificial_intelligence], [ML](https://en.wikipedia.org/wiki/Machine_learning)

Reviewer: ```suggestion
[AI](https://en.wikipedia.org/wiki/Artificial_intelligence), [ML](https://en.wikipedia.org/wiki/Machine_learning)
```

Author: :eyes:

  Pods.

### The new implementation

When a controller needs to take an action on objects before they are removed, it
should add a [finalizer](/docs/concepts/overview/working-with-objects/finalizers/)
to the objects that it manages.
A finalizer prevents the objects from being deleted from the API until the
finalizers are removed. Once the controller is done with the cleanup and
accounting for the deleted object, it can remove the finalizer from the object and the
control plane removes the object from the API.

This is what the new Job controller is doing: adding a finalizer during Pod
creation, and removing the finalizer after the Pod has terminated and has been
accounted for in the Job status. However, it wasn't that simple.

The main challenge is that there are at least two objects involved: the Pod
and the Job. While the finalizer lives in the Pod object, the accounting lives
in the Job object. There is no mechanism to atomically remove the finalizer in
the Pod and update the counters in the Job status. Additionally, there could be
more than one terminated Pod at a given time.

To solve this problem, we implemented a three staged approach, each translating
to an API call.

> 1. Add the terminated Pod(s) UID into temporary lists in the Job status

Reviewer: How about:
```suggestion
1. For each terminated Pod, add the unique ID (UID) of the terminated Pod into a short-lived
   list, stored within the `.status` of the owning Job
```

   ([.status.uncountedTerminatedPods](/docs/reference/kubernetes-api/workload-resources/job-v1/#JobStatus)).
2. Remove the finalizer from the Pods(s).

> 3. Atomically remove UIDs from the temporary lists and increment the `succeeded`
>    and `failed` counters in the Job status.

Reviewer: ```suggestion
3. Atomically remove UIDs from the temporary lists; at the same time, increment the overall
   `succeeded` and `failed` counters in the `status` of the Job.
```
?

Author: how about a bullet list?

Reviewer: Bullets are fine by me.

Additional complications come from the fact that the Job controller might
receive the results of the API changes in steps 1 and 2 out of order. We solved
this by adding an in-memory cache for removed finalizers.

Still, we faced some issues during the beta stage, leaving some pods stuck
with finalizers in some conditions ([#108645](https://github.com/kubernetes/kubernetes/issues/108645),
[#109485](https://github.com/kubernetes/kubernetes/issues/109485), and
[#111646](https://github.com/kubernetes/kubernetes/pull/111646)). As a result,

> we decided to disable the feature in the 1.23 and 1.24 releases.

Reviewer: (nit)
```suggestion
we decided switch that feature gate to be disabled by default for the 1.23 and 1.24 releases.
```

Author: Done

Once resolved, we re-enabled the feature for the 1.25 release. Since then, we
have received reports from our customers running tens of thousands of Pods at a

> time in their clusters through the Job API. With this success, we decided to

Reviewer: ```suggestion
time in their clusters through the Job API. Seeing this success, we decided to
```

Author: Done

[#109485](https://github.com/kubernetes/kubernetes/issues/109485), and
[#111646](https://github.com/kubernetes/kubernetes/pull/111646)). As a result,
we decided to switch that feature gate to be disabled by default for the 1.23
and 1.24 releases.

Once resolved, we re-enabled the feature for the 1.25 release. Since then, we
have received reports from our customers running tens of thousands of Pods at a
time in their clusters through the Job API. Seeing this success, we decided to
graduate the feature to stable in 1.26, as part of our long term commitment to
make the Job API the best way to run large batch Jobs in a Kubernetes cluster.

To learn more about the feature, you can read the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods).

## Acknowledgments

As with any Kubernetes feature, multiple people contributed to getting this
done, from testing and filing bugs to reviewing code.

On behalf of SIG Apps, I would like to especially thank Jordan Liggitt (Google)
for helping me debug and brainstorm solutions for more than one race condition

> and Maciej Szulik (Red Hat) for his conscious reviews.

Author: conscious: having knowledge of something; aware.

But in any case #38470
